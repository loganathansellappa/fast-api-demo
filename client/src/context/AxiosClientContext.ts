import axios from 'axios'
import { createContext, useContext } from 'react'

const Context = createContext<AxiosClientService | undefined>(undefined)
export const AxiosClientProvider = Context.Provider

export const useAxiosClient = () => {
  const context = useContext(Context)
  if (context === undefined) {
    throw new Error('useAxiosClient must be used within a AxiosClientProvider')
  }
  return context
}

export class AxiosClientService {
  public axiosClient: ReturnType<typeof axios.create>
  private endpoint: string

  constructor({ endpoint }: { endpoint?: string; token?: string }) {
    this.endpoint = `${endpoint}`

    this.axiosClient = axios.create({
      baseURL: this.endpoint,
    })
  }
}
