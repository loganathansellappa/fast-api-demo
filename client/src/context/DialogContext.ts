import { Dispatch } from 'react';
import { createContext, SetStateAction, useContext } from 'react';

export type DialogState = {
  isOpen: boolean;
  content: React.ReactNode | null;
  handler: () => void;
  title: string;
  noBtnText: string;
  yesBtnText: string;
};

export type DialogContextType = {
  dialog: DialogState;
  setDialog: Dispatch<SetStateAction<DialogState>>;
};

export const DialogContext = createContext<DialogContextType | null>(null);


export const useDialogContext = () => {
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  const context: any = useContext(DialogContext)
  if (context === undefined) {
    throw new Error('useAxiosClient must be used within a AxiosClientProvider')
  }
  return context;
}