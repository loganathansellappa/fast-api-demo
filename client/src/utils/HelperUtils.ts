import { toast } from 'react-toastify';

export type ServerConfig = {
    baseUrl: string;
};
export const serverData = () => {
    return {
        baseUrl: process.env.DATA_URL,
    } as ServerConfig;
};

export enum MissionState {
    PreFlight = 'Pre-flight',
    Flight = 'In-Flight',
    PostFlight = 'Post-Flight'
}


export enum MutationType {
    PUT = 'PUT',
    POST = 'POST',
    DELETE = 'DELETE'
}

export const showSuccessToast = (message: string, id: string | number = 1) => {
    toast.dismiss();
    toast.success(message, { toastId: id, autoClose: 500, style: { background: '#5CDD00', color: "white" } });
};

export const showErrorToast = (message: string, id: string | number) => {
    toast.dismiss();
    toast.error(message, { toastId: id, autoClose: 500 });
};
