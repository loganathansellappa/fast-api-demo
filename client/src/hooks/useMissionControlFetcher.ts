import {useAxiosClient} from "../context/AxiosClientContext.ts";
import {useMemo} from "react";
import {missionControlKeys} from "../services/queryKeyFactory.ts";
import {useQuery} from "react-query";
import { MissionControlSchema, NormalizedMissionControlSchema } from '../@types/zodSchema.ts';

type TransformedMissionControlData = {
    [key: string]: MissionControlSchema[]
}
export function useMissionControlFetcher() {
    const axiosClientService = useAxiosClient()

    const { isLoading, isError, error, data } = useQuery({
        queryKey: missionControlKeys.all(),
        queryFn: () => axiosClientService.axiosClient.request({ url: 'flight_missions?skip=0&limit=100' }),
    })

    const memoedData = useMemo(() => {
        if (data && data.data) {
            const validData = NormalizedMissionControlSchema.parse(data.data);
            const groupedData: { [key: string]: MissionControlSchema[] } =  validData.reduce((acc: TransformedMissionControlData, obj: MissionControlSchema) => {
                const key = obj.mission_state;
                if (!acc[key]) {
                    acc[key] = [];
                }
                acc[key].push(obj);
                return acc;
            }, {})
            return groupedData
        }

        return {}

    }, [data])

    return {
        data: memoedData,
        isLoading,
        isError,
        error,
    }
}
