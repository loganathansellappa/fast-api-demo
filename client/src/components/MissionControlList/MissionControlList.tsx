import { useMissionControlFetcher } from '../../hooks/useMissionControlFetcher.ts';
import { MissionControlCardContainer } from '../MissionControlCardContainer/MissionControlCardContainer.tsx';
import { MissionState } from '../../utils/HelperUtils.ts';
import { useCallback } from 'react';
import { ContextMenuProvider } from '../ContextMenu/ContextMenuProvider.tsx';
export const MissionControlList = () => {
  const { data, isLoading } = useMissionControlFetcher();

  const containerData = useCallback(() => {
    return Object.values(MissionState).map(state => {
      data[state] = data[state] || []
      const options = Object.values(MissionState).filter(title => title !== state);
      return (
        <div className="shadow-md flex-col w-[40%] m-10 bg-gray-100 p-8 max-2xl no-scrollbar overflow-y-scroll">
          <ContextMenuProvider options={options} key={`ctxProvider-${state}`}>
            <MissionControlCardContainer key={state} missionControl={data[state]} title={state} />
          </ContextMenuProvider>
        </div>
        )
    })
  }, [data])


  if (!data || isLoading) {
    return <div>Loading...</div>;
  }

  return (
      <div className="flex flex-row gap-1 h-full py-10 justify-items-center">
        {containerData()}
      </div>
  );
};



