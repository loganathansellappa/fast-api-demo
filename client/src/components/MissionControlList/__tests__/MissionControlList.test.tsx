import { render } from "@testing-library/react";
import { MissionState } from '../../../utils/HelperUtils.ts';
import * as Hooks from '../../../hooks/useMissionControlFetcher.ts';
import { MissionControlList } from '../MissionControlList.tsx';
import { AxiosClientProvider, AxiosClientService } from '../../../context/AxiosClientContext.ts';
import { QueryClient, QueryClientProvider } from 'react-query';
import { createContext } from 'react';
import * as Contexts from '../../../context/DialogContext.ts';

let element: { container: HTMLElement };


beforeEach(() => {

  const context = createContext<Contexts.DialogContextType>({
    dialog: {
      isOpen: false,
      content: null,
      handler: () => jest.fn(),
      title: "",
      noBtnText: "",
      yesBtnText: "",
    },
    setDialog: () => {}
  })
  jest
    .spyOn(Contexts, "useDialogContext")
    .mockReturnValue(context);
  jest.spyOn(Hooks, "useMissionControlFetcher").mockReturnValue({
    data: {
      [MissionState.PreFlight]: [
        {
          "title": "string",
          "description": "string",
          "mission_state": "Pre-flight",
          "id": 319,
          "created_at": "2024-02-14T00:45:50",
          "updated_at": "2024-02-14T00:54:26"
        },
        {
          "title": "new",
          "description": "ew",
          "mission_state": "Pre-flight",
          "id": 327,
          "created_at": "2024-02-14T00:52:04",
          "updated_at": "2024-02-14T00:52:04"
        },
      ],
      [MissionState.Flight]: [
        {
          "title": "dsa",
          "description": "dsa",
          "mission_state": "In-Flight",
          "id": 328,
          "created_at": "2024-02-14T00:54:12",
          "updated_at": "2024-02-14T00:54:28"
        }
      ],
    },
    isLoading: false,
    isError: false,
    error: false,
  });

  element = render(
    <AxiosClientProvider value={new AxiosClientService({ endpoint: 'url' })}>
      <QueryClientProvider client={new QueryClient()}>
        <MissionControlList />
      </QueryClientProvider>
    </AxiosClientProvider>
  );
});

describe("MissionControlList", () => {
  test("renders MissionControlList component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });
});
