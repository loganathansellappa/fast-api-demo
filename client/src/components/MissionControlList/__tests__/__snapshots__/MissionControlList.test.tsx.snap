// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionControlList renders MissionControlList component 1`] = `
<div
  class="mission-list flex flex-row gap-1 h-full py-10 justify-items-center"
>
  <div
    class="shadow-md flex-col w-[40%] m-10 bg-gray-100 p-8 max-2xl no-scrollbar overflow-y-scroll"
  >
    <div>
      <div
        class="bg-gray-100 text-black font-bold mb-12 rounded sticky"
      >
        Pre-flight
      </div>
      <div
        class="no-scrollbar overflow-y-scroll "
      >
        <div
          class="relative bg-white shadow-md rounded-md p-4 mb-4 min-w-[40%] border-2 hover:cursor-pointer orange-border"
          data-item-id="319"
        >
          <div
            class="flex border-b-2 pb-2"
            data-item-id="319"
          >
            <div
              class="text-lg font-semibold"
              data-item-id="319"
            >
              string
            </div>
            <div
              class="ml-auto"
              data-item-id="319"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-gray-600 mt-4"
            data-item-id="319"
          >
            string
          </div>
        </div>
        <div
          class="relative bg-white shadow-md rounded-md p-4 mb-4 min-w-[40%] border-2 hover:cursor-pointer orange-border"
          data-item-id="327"
        >
          <div
            class="flex border-b-2 pb-2"
            data-item-id="327"
          >
            <div
              class="text-lg font-semibold"
              data-item-id="327"
            >
              new
            </div>
            <div
              class="ml-auto"
              data-item-id="327"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-gray-600 mt-4"
            data-item-id="327"
          >
            ew
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="shadow-md flex-col w-[40%] m-10 bg-gray-100 p-8 max-2xl no-scrollbar overflow-y-scroll"
  >
    <div>
      <div
        class="bg-gray-100 text-black font-bold mb-12 rounded sticky"
      >
        In-Flight
      </div>
      <div
        class="no-scrollbar overflow-y-scroll "
      >
        <div
          class="relative bg-white shadow-md rounded-md p-4 mb-4 min-w-[40%] border-2 hover:cursor-pointer blue-border"
          data-item-id="328"
        >
          <div
            class="flex border-b-2 pb-2"
            data-item-id="328"
          >
            <div
              class="text-lg font-semibold"
              data-item-id="328"
            >
              dsa
            </div>
            <div
              class="ml-auto"
              data-item-id="328"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
                />
              </svg>
            </div>
          </div>
          <div
            class="text-gray-600 mt-4"
            data-item-id="328"
          >
            dsa
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="shadow-md flex-col w-[40%] m-10 bg-gray-100 p-8 max-2xl no-scrollbar overflow-y-scroll"
  >
    <div>
      <div
        class="bg-gray-100 text-black font-bold mb-12 rounded sticky"
      >
        Post-Flight
      </div>
      <div
        class="no-scrollbar overflow-y-scroll "
      />
    </div>
  </div>
</div>
`;
